### DISCUSSION TITLE
how do i recreate Sonic CD's camera extension in this engine?

### DISCUSSION THREAD

**Discussion by thefacer** [Techokami/SonicWorldsNext#172] [Q&A] [answered]
whenever sonic runs or prepares a peelout/spindash, the camera pans in the direction he's going, like this:

![Qqgoodfuture](https://github.com/user-attachments/assets/558797ad-0ee2-4cf0-9714-9eed9ce99194)

what values would i have to edit or add to make it work? all i know is that i have to add a variable that stores the camera pan

**Comment by thefacer** [comment]
**UPDATE:** i managed to find a way on how to make it (partially) work through the help of AI. it only happens if you're pressing a direction and NOT if you're running, though. it doesn't accommodate for camera boundaries either. i'm hoping someone with more experience could make it better

within the _Player.gd_ file, add three variables (maxPanSpeed makes sure it doesn't pan out further than it does):

```
# used for camera panning
var panSpeed = 50.00
var panSmooth := 0.1
var maxPanSpeed = panSpeed
```

then, in the if statement that checks whether the player is pressing the direction they're moving in, you include this:

```
if abs(movement.x) < 10*60:
		var targetOffset = float(direction) * float(panSpeed)
		var newOffset = lerp(float(camera.offset.x), targetOffset, 0.1)
							
					if abs(newOffset) <= maxPanSpeed:
					camera.offset.x = newOffset
```

when you come to a stop if neither of the keys are pressed, add this else statement (to make sure it doesn't lock in that position):

```
else:
	camera.offset.x = lerp(camera.offset.x, 0.0, 0.1)
```

and that's how you get the camera to extend in Sonic Worlds Next! sort of :)

**Reply by fishingrosd113** [reply]
Could you post an image of the code in godot itself? It doesn't seem to be working for me

**Reply by NeonSRB2** [reply, marked as answer]
this code is kinda wierd, so this is a different take without ai
sonic cd's extended camera replaces the horizontal drag margins used in the other classic sonic games.
This is the place to disable that:
```
	# Camera settings
	get_parent().call_deferred("add_child", (camera))
	camera.enabled = (playerControl == 1)
	var viewSize = get_viewport_rect().size
	camera.drag_left_margin =   camDist.x/viewSize.x # make this 0
	camera.drag_right_margin =  camDist.x/viewSize.x # this too
	camera.drag_top_margin =    camDist.y/viewSize.y
	camera.drag_bottom_margin = camDist.y/viewSize.y
```
next, is the actual offsetting.
Sonic Worlds Next uses a variable called `camAdjust` to offset where the camera thinks the center of the player is, because the player actually moves downward a little when rolling. This is a Vector2, but the x component is unused. It is set to 0 a couple times though, so let's get rid of that.
Starting at ~line 1271:
```
	# Extra drag margin for rolling
	match(character):
		CHARACTERS.TAILS:
			match($HitBox.shape.size):
				currentHitbox.ROLL:
					camAdjust.y = -1
				_:
					camAdjust.y = 0
		_: # default
			match($HitBox.shape.size):
				currentHitbox.ROLL:
					camAdjust.y = -5
				_:
					camAdjust.y = 0
```
Now the x-axis is completely untouched. The Sonic Physics Guide on Sonic Retro mentions that the camera moves at a rate of 2 pixels per frame, and shifts when running at at least 6 pixels per frame in either direction, or by doing a spindash/peelout.
with that knowledge, include this afterwards:
```
	# Extended Camera
	if currentState == STATES.SPINDASH or currentState == STATES.PEELOUT: # charging a spindash or peelout
		camAdjust.x += direction*2
	elif movement.x >= 6*60: # running right
		camAdjust.x += 2 # move right
	elif movement.x <= -6*60: # running left
		camAdjust.x -= 2
	elif camAdjust.x != 0: # the check for 0 is here to prevent it moving past 0
		camAdjust.x -= sign(camAdjust.x)*2 # move back to center if not any of this
	camAdjust.x = clamp(camAdjust.x, -64, 64) # stop at 64 pixels
```
I know it's a lot of elifs, but I don't know if I can use match in this situation.
Works fine on my end, so all you should have to do is make sure you remove your old implementation.

**Reply by fishingrosd113** [reply]
Thank you! I think it's very faithful so you did a good job
